File: Lecture37/vmc.cpp
// Variational Monte Carlo for the harmonic oscillator

#include <cmath>
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <vector>
#include "random.hpp"



class QHO {
public :

  QHO(int Nin, double alphain, int MCStepsin) : 
    N(Nin), alpha(alphain), MCSteps(MCStepsin)
  {
    rng.set_seed(123456);

    x.resize(N);
    for (int i = 0; i < N; i++)
      x[i] = rng.rand() - 0.5;
    delta = 1;

    nPsiSqd = int((xMax - xMin) / dx);
    psiSqd.resize(nPsiSqd);
    nAccept = 0;

    zeroAccumulators();
  }

  void zeroAccumulators() {
    eSum = eSqdSum = 0;
    for (int i = 0; i < nPsiSqd; i++)
      psiSqd[i] = 0;
  }

  double p(double xTrial, double x) {

    // compute the ratio of rho(xTrial) / rho(x)
    return exp(- 2 * alpha * (xTrial*xTrial - x*x));
  }

  double eLocal(double x) {

    // compute the local energy
    return alpha + x * x * (0.5 - 2 * alpha * alpha);
  }

  void MetropolisStep() {

    // choose a walker at random
    int n = int(rng.rand() * N);

    // make a trial move
    double xTrial = x[n] + delta * rng.rand_gauss();

    // Metropolis test
    if (p(xTrial, x[n]) > rng.rand()) {
      x[n] = xTrial;
      ++nAccept;
    }

    // accumulate energy and wave function
    double e = eLocal(x[n]);
    eSum += e;
    eSqdSum += e * e;
    int i = int((x[n] - xMin) / dx);
    if (i >= 0 && i < nPsiSqd)
      psiSqd[i] += 1;
  }

  void oneMonteCarloStep() {

    // perform N Metropolis steps
    for (int i = 0; i < N; i++) {
      MetropolisStep();
    }
  }

  void adjustStep() {
    // perform 20% of MCSteps as thermalization steps
    // and adjust step size so acceptance ratio ~50%
    int thermSteps = int(0.2 * MCSteps);
    int adjustInterval = int(0.1 * thermSteps) + 1;

    std::cout << " Performing " << thermSteps << " thermalization steps ..."
	 << std::flush;
    for (int i = 0; i < thermSteps; i++) {
      oneMonteCarloStep();
      if ((i+1) % adjustInterval == 0) {
	delta *= nAccept / (0.5 * N * adjustInterval) ;
	nAccept = 0;
      }
    }
    std::cout << "\n Adjusted Gaussian step size = " << delta << std::endl;    
  }


  // production steps
  void doProductionSteps( ) {
    zeroAccumulators();
    nAccept = 0;
    std::cout << " Performing " << MCSteps << " production steps ..." << std::flush;
    for (int i = 0; i < MCSteps; i++)
      oneMonteCarloStep();
  }

  void print() {

    // compute and print energy
    double eAve = eSum / double(N) / MCSteps;
    double eVar = eSqdSum / double(N) / MCSteps - eAve * eAve;
    double error = sqrt(eVar) / sqrt(double(N) * MCSteps);
    std::cout << "\n <Energy> = " << eAve << " +/- " << error
	      << "\n Variance = " << eVar << std::endl;

    // write wave function squared in file
    std::ofstream file("psiSqd.data");
    double psiNorm = 0;
    for (int i = 0; i < nPsiSqd; i++)
      psiNorm += psiSqd[i] * dx;
    for (int i = 0; i < nPsiSqd; i++) {
      double x = xMin + i * dx;
      file << x << '\t' << psiSqd[i] / psiNorm << '\n';
    }
    file.close();
    std::cout << " Probability density written in file psiSqd.data" << std::endl;    
  }
  

protected :

  int N;                         // number of walkers
  std::vector<double> x;         // walker positions
  double delta;                  // step size
  double eSum;                   // accumulator to find energy
  double eSqdSum;                // accumulator to find fluctuations in E
  double xMin = -10;             // minimum x for histogramming psi^2(x)
  double xMax = +10;             // maximum x for histogramming psi^2(x)
  double dx = 0.1;               // psi^2(x) histogram step size
  std::vector<double> psiSqd;    // psi^2(x) histogram
  int nPsiSqd;                   // size of array
  double alpha;                  // trial function is exp(-alpha*x^2)
  int nAccept;                   // accumulator for number of accepted steps
  cpt::Random rng;               // random number generator
  int MCSteps;                   // number of MC steps
};
  
int main() {

  int N; double alpha; int MCSteps;

  std::cout << " Variational Monte Carlo for Harmonic Oscillator\n"
       << " -----------------------------------------------\n";
  std::cout << " Enter number of walkers: ";
  std::cin >> N;
  std::cout << " Enter parameter alpha: ";
  std::cin >> alpha;
  std::cout << " Enter number of Monte Carlo steps: ";
  std::cin >> MCSteps;

  
  QHO qho(N,alpha,MCSteps);
  
  qho.adjustStep();

  qho.doProductionSteps();
  
  qho.print();

}

File: Lecture38/vmc.py
# Variational Monte Carlo for the harmonic oscillator

import math
import random

def zero_accumulators():
    global psi_sqd, e_sum, e_sqd_sum
    e_sum = e_sqd_sum = 0.0
    for i in range(n_psi_sqd):
        psi_sqd[i] = 0.0

def initialize():
    global x, n_psi_sqd, psi_sqd
    x = []
    for i in range(N):
        x.append(random.random() - 0.5)
    n_psi_sqd = int((x_max - x_min) / dx)
    psi_sqd = [ 0.0 ] * n_psi_sqd
    zero_accumulators()

# the trial function is exp(-alpha x^2)

def p(x_trial, x_i):
    # compute ratio of rho(x_trial) / rho(x_i)
    return math.exp( - 2 * alpha * (x_trial**2 - x_i**2))

def e_local(x_i):
    # compute the local energy
    return alpha + x_i**2 * (0.5 - alpha**2)

def Metropolis_step():
    global x, n_accept, e_sum, e_sqd_sum, psi_sqd
    # choose a walker at random
    i = random.randrange(N)
    # make a trial move
    x_trial = x[i] + delta * random.normalvariate(0.0, 1.0)
    # Metropolis test
    if p(x_trial, x[i]) > random.random():
        x[i] = x_trial
        n_accept += 1
    # accumulate energy and wave function
    e = e_local(x[i])
    e_sum += e
    e_sqd_sum += e**2
    bin = int((x[i] - x_min) / dx)
    if bin >= 0 and bin < n_psi_sqd:
        psi_sqd[bin] += 1.0

def one_Monte_Carlo_step():
    # perform N Metropolis steps
    for i in range(N):
        Metropolis_step()

print(" Variational Monte Carlo for the Harmonic Oscillator")
print(" ---------------------------------------------------")
N = int(input(" Enter number of walkers: "))
alpha = float(input(" Enter parameter alpha: "))
MC_steps = int(input(" Enter number of Monte Carlo steps: "))

x_max = 10.0        # maximum x for histogram
x_min = -x_max
dx = 0.1            # histogram bin width
delta = 1.0         # Metropolis step size

initialize()

# perform 20% of MC_steps as thermalization steps
# and adjust step size so acceptance ratio ~50%
therm_steps = int(0.2 * MC_steps)
adjust_interval = int(0.1 * therm_steps)
n_accept = 0
print(" Performing", therm_steps, "thermalization steps ...")
for i in range(therm_steps):
    one_Monte_Carlo_step()
    if (i + 1) % adjust_interval == 0:
        delta *= n_accept / (0.5 * N * adjust_interval)
        n_accept = 0
print(" Adjusted Gaussian step size =", delta)

# production steps
zero_accumulators()
n_accept = 0
print(" Performing", MC_steps, "production steps ...")
for i in range(MC_steps):
    one_Monte_Carlo_step()

# compute and print energy
e_ave = e_sum / float(N) / MC_steps
e_var = e_sqd_sum / float(N) / MC_steps - e_ave**2
error = math.sqrt(e_var) / math.sqrt(float(N) * MC_steps)
print(" <Energy> =", e_ave, "+/-", error, "Variance =", e_var)

# write wave function squared in file
file = open("psi_sqd.data", "w")
psi_norm = 0.0
for i in range(n_psi_sqd):
    psi_norm += psi_sqd[i] * dx
for i in range(n_psi_sqd):
    x_i = x_min + i * dx
    file.write(" " + str(x_i) + "\t" + str(psi_sqd[i] / psi_norm) + "\n")
file.close()
print(" Probability density written in file psi_sqd.data")
